Introduce more strict compiler warnings and fix them

Ryex opened the pull request:
Add some reasonable compiler warnings to help prevent future problems.
clean up after them.

Ryex committed:
chore: add better default compiler warnings

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
chore: clean up after new compiler warnings

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
chore: fix test & clean up one last warning

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Scrumplex commented:
This might cause issues with Qt 5.12 < x < 5.15 as we only support these versions specifically. Some options might have been deprecated between 5.12 and 5.15 but only replaced for >=5.15 in our codebase

Ryex commented:
> This might cause issues with Qt 5.12 < x < 5.15 as we only support these versions specifically. Some options might have been deprecated between 5.12 and 5.15 but only replaced for >=5.15 in our codebase

I'm not sure how this would break anything. none of these warning warn about deprecation. and they are only warnings. if they flag something that would fall in that feature gap we know to put it in pre-processor flags or special case it. right?

Scrumplex commented:
> I'm not sure how this would break anything. none of these warning warn about deprecation

Let's say something was deprecated in Qt 5.14 in favor of something else and our code only handles this in the form of a `#if QT_VERSION >= QT_VERSION_CHECK(5, 15, 0)` check, building Prism Launcher against Qt 5.14 would raise a deprecation warning, which would then abort the build as you have added `-Werror` (and MSVC equivalents)

Ryex commented:
> > I'm not sure how this would break anything. none of these warning warn about deprecation
> 
> Let's say something was deprecated in Qt 5.14 in favor of something else and our code only handles this in the form of a `#if QT_VERSION >= QT_VERSION_CHECK(5, 15, 0)` check, building Prism Launcher against Qt 5.14 would raise a deprecation warning, which would then abort the build as you have added `-Werror` (and MSVC equivalents)

`-Werror` and msvc equivalent `/WX` are locked behind the setting `WARNINGS_AS_ERRORS`
who ever is building would have to explicitly set `WARNINGS_AS_ERRORS=yes` for them to be added.

Ryex committed:
Merge branch 'develop' into chore/add-compiler-warnings

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Scrumplex commented:
> `-Werror` and msvc equivalent `/WX` are locked behind the setting `WARNINGS_AS_ERRORS`
> who ever is building would have to explicitly set `WARNINGS_AS_ERRORS=yes` for them to be added.

Most packagers would define these kinds of flags using `CMAKE_CXX_FLAGS`. Many build tools (Nix, Makepkg, Flatpak-Builder) set these variables (sometimes indirectly) anyway, so I don't think we *need* this if we don't plan to make it the default.

Ryex committed:
chore: remove WARNINGS_AS_ERRORS

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
chore: fix propagate mispelling

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
chore: fix shadowed member and signed/unsigned mismatch

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

chore: supress unused with [[maybe_unused]]

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

chore: unshadow ^&^& static_cast implicit return

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

chore: deshadow and mark unused in parse task

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

chore: mark unused in folder models

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

chore: deshadow and mark unused with instances

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

chore: more deshadow and unused

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

chore: remove uneeded simicolons

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

chore: mark unused

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

chore: prevent shadow

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
fix: toml without exceptions usage

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
feat: ensure MSVC using stack unwind semantics

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex commented:
hoo

I just cleaned up all the warnings I get from MSVC

I think we might want to add [ac89ac3](https://github.com/PrismLauncher/PrismLauncher/pull/1107/commits/ac89ac37f03571fe3f9e0639f1a7377463b60870)  and [cc41b03](https://github.com/PrismLauncher/PrismLauncher/pull/1107/commits/cc41b039e6ba2a24c0ded9fb573bee6050aa2ac9) to our blame ignore
when this is merged but this really is ready I think.

I've removed the WARNINGS_AS_ERRORS option

Ryex committed:
Merge branch 'develop' into chore/add-compiler-warnings

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
Merge branch 'develop' into chore/add-compiler-warnings

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
chore(rename): fix unfinshed renames from deshadow changes

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Scrumplex commented on launcher/minecraft/PackProfile.cpp:
```diff
@@ -532,9 +532,9 @@
+ComponentPtr PackProfile::getComponent(size_t index)
 {
-    if(index < 0 || index >= d->components.size())
+    if(index < 0 || index >= static_cast<size_t>(d->components.size()))
```
hmm size_t is unsigned I think?

Ryex replied:
the explicit cast was to suppress a warning about data loss in comparing different types.
so yes, if `size_t` unsigned `index < 0` would always be false.

Scrumplex replied:
Yes. But you changed the function signature to use size_t so the `index < 0` doesn't do anything

Ryex committed:
fix(compiler warnings): suppress omtimisation flag conflict warnings. leave optimisation to Byild type defaults
- enable ASan for RelWithDebInfo builds

Ryex commented on launcher/minecraft/PackProfile.cpp:
```diff
@@ -533,9 +533,9 @@ ComponentPtr PackProfile::getComponent(const QString &id)
+ComponentPtr PackProfile::getComponent(size_t index)
 {
-    if(index < 0 || index >= d->components.size())
+    if(index < 0 || index >= static_cast<size_t>(d->components.size()))
```
```suggestion
    if(index >= static_cast<size_t>(d->components.size()))
```

Ryex committed:
Fix: always false comparison

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
DCO Remediation Commit for Rachel Powers <508861+Ryex@users.noreply.github.com>

I, Rachel Powers <508861+Ryex@users.noreply.github.com>, hereby add my Signed-off-by to this commit: b4ae6e9774c535bf02b06325e6763d2d7ee02f21

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Scrumplex committed:
Merge remote-tracking branch 'upstream/staging' into chore/add-compiler-warnings

Scrumplex committed:
fix: fix more warnings

Signed-off-by: Sefa Eyeoglu <contact@scrumplex.net>

Scrumplex commented:
Oh. Clangd added a bad include there oops.

getchoo commented on CMakeLists.txt:
```diff
@@ -85,38 +92,40 @@ set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -DTOML_ENABLE_FLOAT16=0")
 
 # If this is a Debug build turn on address sanitiser
-if (CMAKE_BUILD_TYPE STREQUAL "Debug" AND DEBUG_ADDRESS_SANITIZER)
+if ((CMAKE_BUILD_TYPE STREQUAL "Debug" OR CMAKE_BUILD_TYPE STREQUAL "RelWithDebInfo") AND DEBUG_ADDRESS_SANITIZER)
```
everything else looks good, but i'm really not sure about this...address sanitizer is great, but does have some overhead associated with it; and it's not exactly uncommon for `RelWithDebInfo` to be used by some linux distributions or even users. imo, a better option would be to use asan by default on debug builds, while keeping it as an option for `RelWithDebInfo` or even regular `Release` builds

Trial97 replied:
I do believe that address sanitizer should not be on by default for any kind of build, but instead, explicitly activate them where needed.

getchoo replied:
yeah this would make sense after https://github.com/PrismLauncher/PrismLauncher/pull/1518

txtsd replied:
@getchoo could you hit Resolve so we can merge?

Scrumplex committed:
Merge remote-tracking branch 'upstream/staging' into chore/add-compiler-warnings
